python/ray/ml/train/data_parallel_trainer.py

```diff
@@ -181,4 +201,132 @@ def __init__(
+        if num_params > 1:
+            raise ValueError(
+                f"train_loop_per_worker should take in 0 or 1 arguments, "
+                f"but it accepts {num_params} arguments instead."
+            )
+
+        backend_config = backend_config if backend_config else BackendConfig()
+        self.backend_config = backend_config
+
+        self.train_env_var_values = {
+            var_name: os.environ[var_name]
+            for var_name in _BACKEND_ENV_VARS
+            if var_name in os.environ
+        }
+
+    def training_loop(self) -> None:
+        scaling_config_dataclass = ScalingConfigDataClass(**self.scaling_config)
+
+        train_loop_per_worker = construct_train_func(
+            self.train_loop_per_worker,
+            self.train_loop_config,
+            fn_arg_name="train_loop_per_worker",
+        )
+
+        runtime_env = {"env_vars": self.train_env_var_values}
+
+        remote_executor = ray.remote(num_cpus=0)(BackendExecutor)
+        additional_resources_per_worker = (
+            scaling_config_dataclass.additional_resources_per_worker
+        )
+
+        backend_executor_actor = remote_executor.options(
+            runtime_env=runtime_env
+        ).remote(
+            backend_config=self.backend_config,
+            num_workers=scaling_config_dataclass.num_workers,
+            num_cpus_per_worker=scaling_config_dataclass.num_cpus_per_worker,
+            num_gpus_per_worker=scaling_config_dataclass.num_gpus_per_worker,
+            additional_resources_per_worker=additional_resources_per_worker,
+            max_retries=0,
+        )
+
+        checkpoint_manager = MLTuneCheckpointManager()
+        checkpoint_manager.on_init(preprocessor=self.preprocessor)
+
+        # Start the remote actors.
+        ray.get(backend_executor_actor.start.remote(initialization_hook=None))
+
+        if self.resume_from_checkpoint:
+            resume_checkpoint_dict = self.resume_from_checkpoint.to_dict()
+        else:
+            resume_checkpoint_dict = None
+
+        # Tell Ray Train to only shard the train dataset and not the other datasets.
+        # This is purely an implementation detail and users do not need to know about
+        # this.
+        updated_dataset_dict = {}
+        for key, value in self.datasets.items():
+            if key == TRAIN_DATASET_KEY:
+                updated_dataset_dict[key] = value
+            else:
+                # Ray Train will strip out the added string before exposing to users.
+                updated_dataset_dict[key + "_NO-SHARD"] = value
+
+        training_iterator = TrainingIterator(
+            backend_executor_actor=backend_executor_actor,
+            backend_config=self.backend_config,
+            train_func=train_loop_per_worker,
+            dataset=updated_dataset_dict if len(updated_dataset_dict) > 0 else None,
+            checkpoint_manager=checkpoint_manager,
+            checkpoint=resume_checkpoint_dict,
```

Comment:
Why do we need to coerce the checkpoint to a dict? Seems that might be inefficient.

Reply:
Is it because it's sent to a remote worker? Maybe to_object_ref() would be better in that case.

Reply:
Yes that's a good point! So currently Ray Train only works with dicts, but in a future PR, just like Tune, we should update Train to also use the Checkpoint abstraction internally. I will leave a TODO for this!